packages/actor-query-operation-sparql-endpoint/lib/LazyCardinalityIterator.ts

```diff
@@ -0,0 +1,109 @@
+import { AsyncIterator } from 'asynciterator';
+import LinkedList from './LinkedList';
+/**
+  An iterator that maintains an internal buffer of items.
+  This class serves as a base class for other iterators
+  with a typically complex item generation process.
+  @extends module:asynciterator.AsyncIterator
+*/
+export class LazyCardinalityIterator<T> extends AsyncIterator<T> {
+  private _buffer?: LinkedList<T>;
+  private _cardinality?: Promise<number>;
+  private _count = 0;
+  private _buffering = true;
+
+  public constructor(private readonly _source: AsyncIterator<T>) {
+    super();
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    (<InternalSource<T>> _source)[DESTINATION] = this;
+    _source.on('readable', destinationSetReadable);
+    _source.on('end', destinationSetReadable);
+    _source.on('error', destinationEmitError);
+    this.readable = _source.readable;
+  }
+
+  public read(): T | null {
+    if (this._buffer) {
+      if (!this._buffer.empty) {
+        return this._buffer.shift()!;
+      }
+      if (!this._buffering) {
+        this.close();
+      }
+
+      this.readable = false;
+      return null;
+    }
+
+    let item: T | null = null;
+    // eslint-disable-next-line no-cond-assign
+    if (this._source.readable && (item = this._source.read()) !== null) {
+      this._count += 1;
+    } else {
+      this.readable = false;
+      if (this._source.done) {
+        this.close();
+      }
+    }
+
+    return item;
+  }
+
+  public getCardinality(): Promise<number> {
+    if (this._cardinality) {
+      return this._cardinality;
+    }
+
+    if (this._source.done) {
+      this.close();
+    }
+
+    if (this.done) {
+      this._cardinality = Promise.resolve(this._count);
+    } else {
+      this._buffer = new LinkedList();
+      this._cardinality = new Promise(resolve => {
+        // TODO: See if this section should reject on error
```

Comment:
@rubensworks Should we reject this promise on the `error` event. Similarly to what is done in `.toArray` https://github.com/RubenVerborgh/AsyncIterator/blob/a94fbf50dddfe0ee09b6dc608bc68d83340d5c76/asynciterator.ts#L337

Reply:
Yes, good catch, we should definitely do this!

Reply:
One more case I forgot to think about - should we also reject if an error event occured before `getCardinality` was called?

Reply:
Good point, that's indeed something that could occur in cases where the stream is not being used, but only the metadata is of relevance.

Reply:
Why only when the stream is not being used. What if `getCardinality` is called before / after / during the consumption of the stream.

Reply:
Those as well indeed, this case was just the first that popped in my mind :-)